### ISSUE TITLE
After upgrading from 2.2.0 to 2.2.1 on Centos7.3.1611 , Ansible fails  to find handlers using include (referenced from an included task)

### ISSUE THREAD

**Issue by levequej** [ansible/ansible#20603]
##### ISSUE TYPE
 - Bug Report

##### ANSIBLE VERSION
```
ansible --version
ansible 2.2.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### CONFIGURATION
I tried forcing handler_includes_static = True in ansible.cfg without success (even though I already added static:yes to the include task in handlers)

##### OS / ENVIRONMENT
Centos7.3.1611

##### SUMMARY
upgrading from 2.2.0 to 2.2.1 on Centos7  broke notification from indluded handlers

similar to https://github.com/ansible/ansible/issues/13485 , but not exactly the same

My exact same code broke after yum update , when updating from ansible 2.2.0 to ansible 2.2.1 on Centos 7.3.1611

ERROR! The requested handler '*********' was not found in either the main handlers list nor in the listening handlers list

NB : if the handler is a simple command, it works fine. Only handlers that are doing "include" are failing.

NB: I don't know if the fact that the notify is made from  within an included task has anything to do with the problem but this is my scenario.

##### STEPS TO REPRODUCE

with a directory structure as follows
```
myrole/handlers/main.yml
                mypackage_postinstall_handler.yml
myrole/tasks/main.yml
             install_my_package.yml
```

and code like that in *myrole/tasks/main.yml*
```yaml
- include: install_my_package.yml
```
with code like that in *install_my_package*

```yaml
- name: install package
  yum:
    name: "mypackage"
    state: present
  notify: postinstall mypackage
```

 and code like that in *myrole/handlers/main.yml*
```yaml
# I tired adding static: yes , but it made no difference
- name: postinstall mypackage
  include: mypackage_postinstall_handler.yml
  static: yes
````

then it fails  with:
```
ERROR! The requested handler 'postinstall mypackage' was not found in either the main handlers list nor in the listening handlers list
```

Thing is also that because package is installed, handler won't be notified on next run, and the whole play is useless.

Is there any workaround ?  I am afraid to be forced to rewrite all my  handlers with something like register stuff, then  when stuff.changed

##### EXPECTED RESULTS
working handlers
##### ACTUAL RESULTS
handler is not found

**Comment by ryansb** [comment]
@bcoca this looks like an includes issue, but I'm not sure how to test it.

**Comment by nuklea** [comment]
Same issue on 2.2.1. I have `roles/project/handlers/main.yml`:

```yml
- name: Prepare static files
  include: static.yml

- name: Restart uwsgi service
  include: uwsgi.yml
```
and `roles/project/tasks/main.yml`:

```yml
- name: Clone project repository
  git:
    repo: git@bitbucket.org:company/project.git
    dest: "{{ home_dir }}"
    accept_hostkey: yes
  notify:
    - Prepare static files
    - Restart uwsgi 
```

I've run playbook and got:

```
ERROR! The requested handler 'Restart uwsgi service' was not found in either the main handlers list nor in the listening handlers list
```

but `Prepare static files` handler works fine 😳

Ansible 2.2.0 works fine for all handlers.

**Comment by Amoki** [comment]
Same issue with Ubuntu 14.04 and 16.04, no problem with Ansible 2.2.0

**Comment by diego-treitos** [comment]
Same issue here in Ubuntu 16.10 and ansible `2.2.1.0`.

**Comment by fauntleroy18** [comment]
I am also seeing this issue on `ansible 2.2.1.0` and `CentOS Linux release 7.3.1611 (Core)` 


I have a playbook at the top-level of a directory, and a `handlers/main.yml` file with notification handlers. This setup was working fine until the upgrade to 2.2.1 - and it now fails with the message: 

`ERROR! The requested handler 'restart apcupsd' was not found in either the main handlers list nor in the listening handlers list`

**Comment by welchwilmerck** [comment]
Here's the problem:

For a role, when defining a handler with include: to perform multiple tasks (i.e., a handler block), the handler must be the first handler (of any sort) in roles/\<your role\>/handlers/main.yml. Otherwise, it isn't added to the "main handlers list".

See https://github.com/welchwilmerck/ansible-20603

@ryansb - Is this a sufficient test case?

**Comment by dkewley** [comment]
I see this issue in a role handlers/main.yml in Ansible 2.3.2.0 from CentOS 7 Extras. I can confirm the workaround by @welchwilmerck works for me (put the handler with the include at the top of the handlers/main.yml).

I see this issue in the context of a role when using 2.3.2.0 from CentOS 7 Extras. It's resolved by moving the handler with the include: to the top of the handlers file, as per @welchwilmerck.

**Comment by jahickey** [comment]
Ansible 2.4.0 with the **import_tasks** module also has this issue and the workaround by @welchwilmerck does work. But unfortunately, I need to have the **import_task** as the second handler listed in the main.yml file.

Basically I'm setting the hostname, then the first handler sets the /etc/hosts file and then the second handler reboots the machine (by importing my reboot-system role's main.yml file) so that subsequent roles have the hostname and network set properly.

**Comment by m-u-xyz** [comment]
Same issue here, except that the workaround does not work – my handler is first in role/foo/handlers/main.yml, uses a block, but is not found.

My workaround is for the handler to set a flag variable which my "cleanup" role then uses to process the steps that originally were in the block.

**Comment by kmadnani** [comment]
I am facing the same issue while using import_tasks for handlers. This happens only for those handlers where in a handler is called from another handler. Do we have a workaround? Are we planning to fix it in upcoming ansible version?

**Comment by sivel** [comment]
This issue was resolved in a889576d390eeab5137d31bd30821b3310c8d59e and included in the 2.4.4 release.
